src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
+            }
+            else
+            {
+                receiver = Receiver;
+            }
+
+            Task.Run(async () =>
+            {
+                while (true)
+                {
+                    try
+                    {
+                        if (_receiver == null || cancellationToken.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
+                        if (_isSessionsEnabled)
+                        {
+                            try
+                            {
+                                receiver = await sessionClient.AcceptMessageSessionAsync();
+                            }
+                            catch (ServiceBusTimeoutException)
+                            {
+                                // it's expected if the entity is empty, try next time
+                                continue;
+                            }
+                        }
+
+                        IList<Message> messages = messages = await receiver.ReceiveAsync(_serviceBusOptions.BatchOptions.MaxMessageCount, _serviceBusOptions.BatchOptions.OperationTimeout);
+
+                        if (messages != null)
+                        {
+                            Message[] messagesArray = messages.ToArray();
+                            ServiceBusTriggerInput input = new ServiceBusTriggerInput()
+                            {
+                                Messages = messagesArray
+                            };
+                            if (_isSessionsEnabled)
+                            {
+                                input.MessageSession = receiver as IMessageSession;
+                            }
+                            else
+                            {
+                                input.MessageReceiver = receiver as MessageReceiver;
+                            }
+
+                            FunctionResult result = await _triggerExecutor.TryExecuteAsync(GetTriggerFunctionData(messagesArray, input), cancellationToken);
+
+                            if (cancellationToken.IsCancellationRequested)
+                            {
+                                return;
+                            }
+
+                            // Complete batch of messages only if the execution was succesfull
+                            if (result.Succeeded && Receiver != null)
+                            {
+                                await receiver.CompleteAsync(messagesArray.Select(x => x.SystemProperties.LockToken));
+                            }
+                        }
+                    }
+                    catch (ObjectDisposedException)
+                    {
+                        // Ignore as we are stopping the host
+                    }
+                    catch (NullReferenceException)
+                    {
+                        // Ignore as we are stopping the host

```

Comment:
nit: if you get a nullref you don't know it's because the host is stopping. I don't think we should be swallowing nullrefs